### PR TITLE
Refactoring MDNorm

### PR DESCRIPTION
**Pull request by wdzhou**
**Description of work.**

Refactoring MDNorm improve readability. 

**To test:**

Related to #30926.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/MDAlgorithms/src/MDNorm.cpp [hunk 1]
```diff
@@ -1136,13 +1139,165 @@ void MDNorm::cacheDimensionXValues() {
...
+      pos[3] = static_cast<coord_t>(m_Ei - pos[3] * pos[3] / energyToK);
+      // signal = energy distance between two consecutive intersections *solid
+      // angle *PC
+      signal = solid * delta;
+      // [Task 89] bkgdSignal = bkgdSolid * delta
+    }
+
+    // Find the coordiate of the new position after transformation
+    m_transformation.multiplyPoint(pos, posNew);
+    // [Task 89] Is linIndex common to both sample and background?
```
Comment (wdzhou) [inline Framework/MDAlgorithms/src/MDNorm.cpp:1279]: @AndreiSavici : please confirm this!

File: Framework/MDAlgorithms/src/MDNorm.cpp [hunk 2]
```diff
@@ -1136,13 +1139,165 @@ void MDNorm::cacheDimensionXValues() {
...
+                     return static_cast<coord_t>(0.5 * (rhs + lhs));
+                   });
+    signal_t signal;
+    // [Task 89 signal_t bkgdSignal;
+    if (m_diffraction) {
+      // Diffraction
+      // index of the current intersection
+      auto k = static_cast<size_t>(std::distance(intersectionsBegin, it));
+      // signal = integral between two consecutive intersections
+      signal = (yValues[k] - yValues[k - 1]) * solid;
```
Comment (wdzhou) [inline Framework/MDAlgorithms/src/MDNorm.cpp:1264]: @AndreiSavici

File: Framework/MDAlgorithms/src/MDNorm.cpp [hunk 3]
```diff
@@ -1136,13 +1139,165 @@ void MDNorm::cacheDimensionXValues() {
...
+      // [Task 89] signalBkgd = (bkgdValues[k] - bkgdValues[k - 1]) * solidBkgd;
```
Comment (wdzhou) [inline Framework/MDAlgorithms/src/MDNorm.cpp:1265]: There won't be bkgdValues... yValues are from intersection integral with flux.  It is only to with Q values (x values of Worksapce2D flux workspace)

File: Testing/SystemTests/tests/framework/MDNormBackgroundHYSPECTest.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    def runTest(self):
+        config.setFacility('SNS')
+        Load(Filename='HYS_13656-13658',OutputWorkspace='sum')
+        SetGoniometer(Workspace='sum', Axis0='s1,0,1,0,1')
+        GenerateEventsFilter(InputWorkspace='sum',
+                             OutputWorkspace='splboth',
+                             InformationWorkspace='info',
+                             UnitOfTime='Nanoseconds',
+                             LogName='s1',
+                             MaximumLogValue=90,
```
Comment (wdzhou) [inline Testing/SystemTests/tests/framework/MDNormBackgroundHYSPECTest.py:36]: Limit to 10 for test purpose of task 85

File: Framework/MDAlgorithms/src/MDNorm.cpp (hunk 1 above)
Comment (AndreiSavici) [reply Framework/MDAlgorithms/src/MDNorm.cpp:1279]: Yes

File: Framework/MDAlgorithms/src/MDNorm.cpp (hunk 3 above)
Comment (wdzhou) [reply Framework/MDAlgorithms/src/MDNorm.cpp:1265]: signalBkgd = (yValues[k] - yValues[k - 1]) * solidBkgd;

**Comment by wdzhou** [comment]
PR is closed by developer without merging.

**Pull request closed without merging**